data/data_interface.py

```diff
@@ -0,0 +1,184 @@
+        data : pd.DataFrame
+            A dataframe of the data to load. If None, we will attempt to load using file_path
+        file_path : str
+            The path to load your data from. Not required with data is not None.
+        continuous_features : Sequence[str]
+            A sequence of the continous features.
+        ordinal_features : Sequence[str]
+            A sequence of the ordinal features.
+        categorical_features : Sequence[str]
+            A sequence of the categorical/nominal features.
+        immutable_features : Sequence[str]
+            A sequence of the immutable features.
+        target_feature : str
+            The column name of the target feature
+        target_mapping : Any
+            A mapping of the target features to some values
+        scaling_method : str
+            String indicating how you with to scale your data. Options are MaxMin normalization and standardization 
+        encoding_method : str
+            Encoding method for categorical features. Only one-hot for now.
+        """
+        self._continuous_features = continuous_features
+        self._ordinal_features =  ordinal_features
+        self._categorical_features = categorical_features
+        self._immutable_features = immutable_features
+        self._target_feature = target_feature
+        self._target_mapping = target_mapping
+        self._scaling_method = scaling_method
+        self._encoding_method = encoding_method
+        if (file_path is None or file_path == "") and isinstance(data, pd.DataFrame):
+            df = data
+        elif file_path is not None and file_path != "":
+            df = pd.read_csv(file_path)
+        else:
+            raise Exception("No data and file path provided")
+        X_features = df.columns[df.columns != target_feature]
+        self._df_y = df[self._target_feature]
+        self._df_X = df[X_features]
+        self._df_y_init = self._df_y.copy()
+        self._df_X_init = self._df_X.copy()
+        self._X_train = None
+        self._X_test = None
+        self._y_train = None
+        self._y_test = None
+        self._pos_label = pos_label
+    
+    def split_data(self, test_size: float = 0.3) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame ,pd.DataFrame]:
+        """
+        Splits data into X and y and train and test data. Should be used as the last step when setting up data.
+
+        Parameters
+        ----------
+        test_size : float
+            Between 0.0 and 1.0 and represents the proportion of the dataset to include in the test split.
+        Returns
+        -------
+        X_train, X_test, y_train, y_test : Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame ,pd.DataFrame]
+            Dataframes of non-target and target features seperated also split into training and testing data.
+        """
+        self._X_train, self._X_test, self._y_train, self._y_test = train_test_split(self._df_X, self._df_y, test_size=test_size, random_state=42)

```

Comment:
We will likely want to save the dataset indices that correspond to the separate splits. (This can either be done by saving both splits or just saving one and doing a if split_ind / if not split_ind to get both splits).

Reply:
To get the datasplits and preprocess, we'd want to do the following:

let `df` be the pointer to the overall dataset object loaded in from CSV. This includes features, labels, and everything that will then be split into train, test, val splits. 

If doing it from scratch, this is a good and reproducible approach:
1. Produce the overall indices for the original, unmodified data. We assume that the data is loaded in deterministically for the datasets we work with. e.g. `indices`: Sequence[int]; `indices = np.array(len(df))`  # [0, ..., n-1]
2. Set up the random number generator using a random seed. Make sure that random seed is a class attribute.
3. Do not shuffle the dataset itself, but shuffle the indices using a random number generator. indices.shuffle(random_num_gen) --> [3,10,2,...] with length n-1  # an array of number 0 to n-1 in random order
4. Get the train_indices,  train_indices = [0 : np.ceiling(len(df) * train_proportion)]
5. Then extract the respective splits, i.e. `df_train <-- df[train_indices]`, same for test and validation. 
6. Then do all preprocessing on each respective split: normalization, standardization, binary encoding of the label value.

OTHERWISE, please use an existing method for getting random train/test/val splits. If using scikit learn's model_selection.train_test_split(), you'll have to split the train split again to get a validation split. THEN you'd proceed by doing preprocessing on the respective splits (independently).

Reply:
You can make deep copy when making the splits so the original DataFrame is unmodified. DO NOT reset the indices after splitting so the indices in each split will align with their place in the original (unmodified) DataFrame, i.e. `df`.